Rewrite CesiumFlyToController to match improvements in cesium-unreal

azrogers opened the pull request:
Fixes #390 and #334.

The `CesiumFlyToController` had a jump at the end of its flight path, similar to the bug in the Unreal plugin. However, the fix that we used for the Unreal plugin couldn't just be dropped in as the way the two implementations work was quite different. This change rewrites `CesiumFlyToController` to match the implementation of the Unreal equivalent, fixing the jump bug in the process.

csciguy8 commented on Runtime/CesiumFlyToController.cs:
```diff
@@ -222,62 +237,67 @@ private void HandleFlightStep(float deltaTime)
+            {
+                flyPercentage = 1.0;
+            }
+            else if (this._flyToProgressCurve != null && this._flyToProgressCurve.length > 0)
```
I like the length check here. This doesn't exist on the Unreal version, which might be a nice improvement.

csciguy8 reviewed (changes requested):
Code looks good to me when comparing against the latest Unreal implementation.

General functionality look ok, but my testing on unity samples, 04_CesiumSubScenes, showed that for small distances, I can see a small jump between at the end of the animation. (move away from the start a bit, then press 1)

azrogers commented on Runtime/CesiumFlyToController.cs:
```diff
@@ -222,62 +237,67 @@ private void HandleFlightStep(float deltaTime)
-                    1.0);
-            }
+            // Rotate the source direction around the axis by the given percentage of the total angle we're rotating, giving us our current position.
+            double3 rotatedDirection = (float3)(Quaternion.AngleAxis((float)(flyPercentage * this._totalAngle), this._rotationAxis) * this._sourceDirection);
```
Nope! Originally `rotatedDirection` was a `Vector3`, so you need to use `float3` as an in-between to allow casting to a `Vector3`, but it turns out it made more sense as a `double3`. I'll remove this.

azrogers replied:
Nevermind, now I remember why I did it - making `rotatedDirection` a `double3` avoids a cast in the next line, and a `Vector3` has to be cast to a `float3` to be assigned to a `double3`.

azrogers commented:
Updated review! @csciguy8. This one is pretty hefty. It's now blocked by CesiumGS/cesium-native#797 - the cesium-native implementation of `GlobeFlightPath` - as well as #380, which is required to get this project up to date with the latest version of cesium-native. Because of that, it'll probably take a while to get this one merged!

With that being said, I've moved the math behind `CesiumFlyToController` onto the cesium-native side so we can take advantage of glm's math functions to get the precision we need. More info about that side of things is in the cesium-native issue linked in the paragraph above. Then there's the glue in the cesium-unity native plugin to connect things up between C++ and Unity.

Speaking of which, I'd like some input from @kring on that part if possible. `CeisumGlobeFlightPath` isn't supposed to be constructed directly - you're supposed to go through `FromEarthCenteredEarthFixedCoordinates` or `FromLongituteLatitudeHeight`. The reason for this is because the creation of the `GlobeFlightPath` can fail (if the coordinates couldn't be mapped to the WGS84 geodetic surface) and needs to be able to return an empty value (the C++ side uses std::optional for this). But it doesn't look like Reinterop supports making a non-public constructor for a class, meaning you can happily call `new CesiumGlobeFlightPath` and you won't know you've done something wrong until an assertion on the C++ side crashes Unity. How can I solve this? Is there a different paradigm I should use for constructing objects where that construction can fail?

kring commented:
> But it doesn't look like Reinterop supports making a non-public constructor for a class, meaning you can happily call new CesiumGlobeFlightPath and you won't know you've done something wrong until an assertion on the C++ side crashes Unity.

Some confusion here I think. Your `CesiumGlobeFlightPath` does have a public default constructor. The C# compiler has generated it because you haven't explicitly declared any constructors at all. And because you're using it from `ConfigureReinterop.cs`, Reinterop has made it accessible from C++.

If you don't want any public constructors, you need to define a private one (even if it has no implementation). Then the line you added to ConfigureReinterop.cs will be a compiler error, and you'll want to simply remove it. Then I think everything will be working as expected, but let me know if not.

azrogers commented:
Thanks, Kevin! With that, I was able to make the constructor private and get everything working!

kring reviewed (commented):
Looks great to me, just a couple trivial comments.

azrogers commented:
@kring Resolved your comments!

kring commented on native~/Runtime/src/CesiumSimplePlanarEllipsoidCurveImpl.h:
```diff
@@ -0,0 +1,45 @@
+      double additionalHeight) const;
+
+private:
+  std::unique_ptr<CesiumGeospatial::SimplePlanarEllipsoidCurve> _curve;
```
I noticed one more small thing worth fixing. It would be a bit better, I think, to make this a `std::optional` instead of a `std::unique_ptr`. The former avoids a heap allocation and just has nicer semantics overall (for example, it allows the `CesiumSimplePlanarEllipsoidCurveImpl` to be copyable; not that we're likely to copy it).

azrogers commented:
@kring Changed to `std::optional`!

kring commented on native~/Runtime/src/CesiumSimplePlanarEllipsoidCurveImpl.cpp:
```diff
@@ -25,12 +25,12 @@ bool CesiumSimplePlanarEllipsoidCurveImpl::
-  this->_curve = std::make_unique<SimplePlanarEllipsoidCurve>(*flightPath);
-  return this->_curve != nullptr;
+  this->_curve = *flightPath;
+  return this->_curve.has_value();
```
This function (and the one below) can be simplified quite a bit now that you're using `std::optional`. Something like this (not tested):

```cpp
  this->_curve =
      SimplePlanarEllipsoidCurve::fromEarthCenteredEarthFixedCoordinates(
          Ellipsoid::WGS84,
          glm::dvec3(sourceEcef.x, sourceEcef.y, sourceEcef.z),
          glm::dvec3(destinationEcef.x, destinationEcef.y, destinationEcef.z));
  return this->_curve.has_value();
```

kring commented:
Looks great, just one last nitpick! :)

csciguy8 commented:
Looks good @azrogers!.

Checked all remaining comments, retested, and a did a sanity check over all the file diffs once more.

Will merge once CI completes...
